content/2022-02-25-pattern-matching-with-named-fields.md

```diff
@@ -0,0 +1,306 @@
+
+Example for user:
+
+```scala
+object User:
+  def unapply(user: User): Option[
+    (String, Int, String) & 
+    { type Names = ("name", "age", "city") }
+  ] = Some((user.name, user.age, user.city)).asInstanceOf
+```
+
+// TODO: Describe why Names can also be use toplevel.
+
+The reason to pull the names out, instead of keeping them near to their type, is to prevent extractors to accidentally leaking the name.
+
+## Implementation
+
+If a pattern with a name is encountered, the compiler looks up list of provided names and places the trees accordingly.
+
+The list of names either provided by the return type of the unapply method or by the constructor list of the case class.
+
+Example:
+
+```scala
+// a match clause like
+case User(<tree1>, city = <tree2>) => ???
+
+// gets rewritten to:
+case User(
+  <tree1>, // because tree1 is positional pattern
+  _, // because name isn't mentioned
+  <tree2>, // because city is the third parameter of user
+)
+```
+
+This allows us to glace over the details of how pattern matching gets desugared further down the line.
+
+## Drawbacks
+
+### Allow skipping arguments
+
+Whenever a single named argument is used in pattern, the pattern can have fewer arguments than the unapply provides. This is driven by the motivation the make pattern matching extensible.
+But this leads to (arguably small) inconsistency, as pointed out by Lionel Parreaux in the [Scala Contributors Thread](https://contributors.scala-lang.org/t/pattern-matching-with-named-fields/1829/44). This could lead users to use a named pattern, just to skip all parameters.
+
+```scala
+  case User(age = _) => "Just wanted to use the extractor, lol!"
+```
+
+### Reordering of user code
+
+
+```scala
+//TODO: Test this
+object User:
+  class UserExtractor(user: User):
+    def _1 = println("Got name"); user.name
+    def _2 = println("Got age"); user.age
+    def _3 = println("Got city"); user.city
+
+    type Names = ("name", "age", "city")
+// Ask for city, then for name
+val User(city = _, name = _) = user
+// But the execution shows:
+// Got name
+// Got age
+// Got city
+```
+
+## Alternatives
+
+### Without any changes to the language
+
+One alternative way of archiving most objectives, that is doable with current Scala, is to use specialized extractors.
+
+```scala
+object User:
+  object age:
+    def unapply(user: User): Option[Int] =
+      Some(user.age)
+
+user match
+  case User.age(y) => y
+```
+
+Libraries like [Monocle][monocle] could be extended to reduce the boilerplate, but some boilerplate would remain.
+In addition, this breaks the intuitive similarity between construction and deconstruction.
+
+### Alternative desugaring
+
+As the above described desugaring has its drawbacks.Here are some alternatives with other drawbacks, and maybe better trade-offs.
+
+#### Use underscore methods
+
+In the sprit of [name based pattern matching](https://dotty.epfl.ch/docs/reference/changed-features/pattern-matching.html#name-based-match):
```

Comment:
Speaking of name-based pattern matching, the main proposal should better describe how it interacts with name-based extractors.

It should also describe how it interacts with single-argument extractors, which don't get a `Tuple` of the elements.